src/hotspot/share/utilities/elfFile.cpp

```diff
@@ -348,4 +302,1411 @@ ElfStringTable* ElfFile::get_string_table(int index) {
+    case DW_FORM_ref1:
+    case DW_FORM_flag:
+    case DW_FORM_flag_present:
+      if (!_reader.move_position(1)) {
+        return false;
+      }
+      break;
+    case DW_FORM_sdata:
+    case DW_FORM_udata:
+    case DW_FORM_ref_udata:
+      if (!_reader.read_uleb128(&next_qword)) {
+        return false;
+      }
+      break;
+    case DW_FORM_indirect:
+      // Should not be used and therefore is not supported by this parser.
+      return false;
+    case DW_FORM_sec_offset:
+      if (is_DW_AT_stmt_list_attribute) {
+        // DW_AT_stmt_list has the DW_FORM_sec_offset attribute encoding. Store the result in _debug_line_offset.
+        // 4 bytes for 32-bit DWARF.
+        log_develop_trace(dwarf)("    Name: DW_AT_stmt_list, Form: DW_FORM_sec_offset");
+        log_develop_trace(dwarf)("    Reading .debug_line offset from compilation unit at " PTR32_FORMAT, (uint32_t)_reader.get_position());
+        if (!_reader.read_dword(&_debug_line_offset)) {
+          return false;
+        }
+        break;
+      } else {
+        if (!_reader.move_position(4)) {
+          return false;
+        }
+        break;
+      }
+    default:
+      log_develop_info(dwarf)("Unknown DW_FORM_* attribute encoding.");
+      return false;
+  }
+  // Reset the index into the file to the original position where the DebugAbbrev reader stopped reading before calling this method.
+  _reader.reset_to_previous_position();
+  return true;
+}
+
+bool DwarfFile::LineNumberProgram::find_filename_and_line_number(char& filename, const size_t filename_len, int& line) {
+  if (!read_header()) {
+    log_develop_info(dwarf)("Failed to parse the line number program header correctly.");
+    return false;
+  }
+  return run_line_number_program(filename, filename_len, line);
+}
+
+// Parsing header as specified in section 6.2.4 of DWARF 4 spec. We do not read the file_names field, yet.
+bool DwarfFile::LineNumberProgram::read_header() {
+  Elf_Shdr shdr;
+  if (!_dwarf_file->read_section_header(".debug_line", shdr)) {
+    log_develop_info(dwarf)("Failed to read the .debug_line section header.");
+    return false;
+  }
+
+  if (!_reader.set_position(shdr.sh_offset + _debug_line_offset)) {
+    return false;
+  }
+
+  if (!_reader.read_dword(&_header._unit_length) || _header._unit_length == 0xFFFFFFFF) {
+    // For 64-bit DWARF, the first 32-bit value is 0xFFFFFFFF. The current implementation only supports 32-bit DWARF format since GCC
+    // only emits 32-bit DWARF.
+    return false;
+  }
+
+  if (!_reader.read_word(&_header._version) || (_header._version != 3 && _header._version != 4)) {
+    // DWARF 3 uses version 3 and DWARF 4 uses version 4 as specified in Appendix F of the DWARF 3 and 4 spec, respectively. For some
+    // reason, GCC is currently using version 3 as specified in the DWARF 3 spec for the line number program even though GCC should
+    // be using version 4 for DWARF 4 as it emits DWARF 4 by default.
+    return false;
```

Comment:
According to the specification (pg112):

> `version (uhalf)`
> A version number (see Appendix F). This number is specific to the line number information
> and is independent of the DWARF version number.

So this is just fine - actually things may break if the code accepted version 4 here assuming that there are breaking differences.
On the other hand Appendix F mentions that DWARF4 contains .debug_line information in version 4.